qt/aqt/sidebar.py

```diff
@@ -918,95 +1030,132 @@ def _notetype_tree(self, root: SidebarItem) -> None:
-                ),
-            )
+    def _maybe_add_tree_actions(self, menu: QMenu) -> None:
+        def set_expanded(expanded: bool) -> None:
+            for index in self.selectedIndexes():
+                self.setExpanded(index, expanded)
 
-        m.exec_(QCursor.pos())
+        def set_children_expanded(expanded: bool) -> None:
+            for index in self.selectedIndexes():
+                self.setExpanded(index, True)
+                for row in range(self.model().rowCount(index)):
+                    self.setExpanded(self.model().index(row, 0, index), expanded)
 
-    def maybe_add_tree_actions(
-        self, menu: QMenu, item: SidebarItem, parent: QModelIndex
-    ) -> None:
         if self.current_search:
             return
-        if not any(bool(c.children) for c in item.children):
-            return
 
-        def set_children_collapsed(collapsed: bool) -> None:
-            m = self.model()
-            self.setExpanded(parent, True)
-            for row in range(m.rowCount(parent)):
-                idx = m.index(row, 0, parent)
-                self.setExpanded(idx, not collapsed)
+        selected_items = self._selected_items()
+        if not any(item.children for item in selected_items):
+            return
 
         menu.addSeparator()
-        menu.addAction(
-            tr(TR.BROWSING_SIDEBAR_EXPAND_CHILDREN),
-            lambda: set_children_collapsed(False),
-        )
-        menu.addAction(
-            tr(TR.BROWSING_SIDEBAR_COLLAPSE_CHILDREN),
-            lambda: set_children_collapsed(True),
-        )
+        if any(not item.expanded for item in selected_items if item.children):
+            menu.addAction(tr(TR.BROWSING_SIDEBAR_EXPAND), lambda: set_expanded(True))
+        if any(item.expanded for item in selected_items if item.children):
+            menu.addAction(
+                tr(TR.BROWSING_SIDEBAR_COLLAPSE), lambda: set_expanded(False)
+            )
+        if any(
+            not c.expanded for i in selected_items for c in i.children if c.children
+        ):
+            menu.addAction(
+                tr(TR.BROWSING_SIDEBAR_EXPAND_CHILDREN),
+                lambda: set_children_expanded(True),
+            )
+        if any(c.expanded for i in selected_items for c in i.children if c.children):
+            menu.addAction(
+                tr(TR.BROWSING_SIDEBAR_COLLAPSE_CHILDREN),
+                lambda: set_children_expanded(False),
+            )
 
-    def rename_deck(self, item: SidebarItem) -> None:
+    def rename_deck(self, item: SidebarItem, new_name: str) -> None:
         deck = self.mw.col.decks.get(item.id)
         old_name = deck["name"]
-        new_name = getOnlyText(tr(TR.DECKS_NEW_DECK_NAME), default=old_name)
-        new_name = new_name.replace('"', "")
-        if not new_name or new_name == old_name:
-            return
+        new_name = re.sub(
+            re.escape(item.name) + "$", new_name.replace("\\", r"\\"), old_name
+        )
```

Comment:
I had trouble following this - could we split it up and comment - something like the following if I'm understanding correctly?

```python
# escape any backslashes the user has provided
new_name = new_name.replace("\\", r"\\")
# modify the full old name, replacing the last component
new_name = re.sub(
            re.escape(item.name) + "$", new_name, old_name
        )
```

Reply:
And same for _rename_tag(). Maybe we could move it into a function?

Reply:
That was a bit hacky. I made the item constructor store the name prefix separately so we don't need re.